backend/YouTubeV2.Application/Services/UserService.cs

```diff
@@ -67,16 +80,179 @@ public async Task RegisterAsync(RegisterDto registerDto, CancellationToken cance
+            CancellationToken cancellationToken)
+        {
+            var userToEdit = await VerifyAccessAndGetUserToModify(callerID, userQueryID);
+
+            await _updateUserDTOValidator.ValidateAndThrowAsync(updateUserDTO, cancellationToken);
+            await VerifyEmailUniqueAsync(userToEdit);
+            await VerifyNameUniqueAsync(userToEdit);
+
+            userToEdit.UserName = updateUserDTO.nickname;
+            userToEdit.Name = updateUserDTO.name;
+            userToEdit.Surname = updateUserDTO.surname;
+
+            if (updateUserDTO.avatarImage.IsNullOrEmpty())
+                await _blobImageService.DeleteProfilePictureAsync(userToEdit.Id, cancellationToken);
+            else
+                await _blobImageService.UploadProfilePictureAsync(updateUserDTO.avatarImage!, userToEdit.Id, cancellationToken);
+
+            var currentRole = await GetRoleForUserAsync(userToEdit);
+            var newRole = updateUserDTO.userType;
+
+            if (currentRole.Equals(Role.Creator, StringComparison.InvariantCulture)
+                && newRole.Equals(Role.Simple, StringComparison.InvariantCulture))
+                await ProcessSwitchToSimple(userToEdit, cancellationToken);
+
+            await SwitchUserRoleAsync(userToEdit, currentRole, newRole);
+            
+            await _userManager.UpdateAsync(userToEdit);
+        }
+
+        public async Task DeleteAsync(string callerID, string userID, CancellationToken cancellationToken)
+        {
+            var callingUser = await GetByIdAsync(callerID);
+
+            if (!await _userManager.IsInRoleAsync(callingUser!, Role.Administrator))
+                throw new ForbiddenException("Only ADMIN can delete users");
+
+            var user = await GetByIdAsync(userID) ?? throw new BadRequestException("User not found");
+
+            var userRoles = await _userManager.GetRolesAsync(user);
+            var userClaims = await _userManager.GetClaimsAsync(user);
+
+            if (await _videoService.GetVideoCountAsync(user) > 0)
+                throw new BadRequestException("Cannot delete account - there are still some videos uploaded");
+
+            foreach (var role in userRoles)
+            {
+                var roleResult = await _userManager.RemoveFromRoleAsync(user, role);
+                if (!roleResult.Succeeded)
+                    throw new BadRequestException(roleResult.Errors.Select(error => new ErrorResponseDTO(error.Description)));
+            }
+
+            foreach (var claim in userClaims)
+            {
+                var claimResult = await _userManager.RemoveClaimAsync(user, claim);
+                if (!claimResult.Succeeded)
+                    throw new BadRequestException(claimResult.Errors.Select(error => new ErrorResponseDTO(error.Description)));
+            }
+
+            await _subscriptionService.DeleteAllSubscriptionsRelatedToUserAsync(user, cancellationToken);
```

Comment:
i think we could have done this by using rules what should happen when we delete a user like cascade

Reply:
Im not sure if it is worth it to change it now